### PR TITLE
Performance improvements since upgrade to 2.16.6

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Addresses some performance issues noted in the upgrade to 2.16.6:
 * Determination of compatible function types was slow across multiple queries. This has been addressed by standardizing on the `Compatible.qll` library, and ensuring all predicates are appropriately bound.
 * In some cases the `isExcluded` mechanism was causing bad join orders. These have been addressed on a case-by-base basis.
 * `RULE-9-4` has been modified to perform a top-down, rather than bottom-up, detection of pairs of expressions initializing the same element in an array aggregate literal. This reduces the cases significantly.

The performance logs for 2.16.6 still note some increases, particularly in the calculation of standard library features such as the IR. I suggest we re-trigger performance analysis on this PR and see whether they replicate.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `DCL40-C`
     - `RULE-5-1`
     - `RULE-8-3`
     - `RULE-8-4`
     - `MSC39-C`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by lcartey** [comment]
Publishing a re-formatted table of total serialized time (in seconds) for easier review:

Version | x86-linux-c | x86-linux-cpp
-- | -- | --
v2.35.0 | 5283.229 | 5283.229
v2.36.0 | 2707.846 | 4154.858
#781 | 2600.1 | 3731.22

With these changes performance is slightly improved for C (4% decrease) and a more significant improvement for C++ (10% decrease) vs. the last release (and a significant performance improvement on both since v2.35.0).

@knewbury01 this should address the slow performance we saw in the 2.37.0 release candidate.

File: c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql [hunk 1]
```diff
@@ -60,35 +60,55 @@ int getMaxDepth(ArrayAggregateLiteral al) {
...
+  or
+  exists(ArrayAggregateLiteral parent1, ArrayAggregateLiteral parent2, int shared_index |
+    hasMultipleInitializerExprsForSameIndexInternal(root, parent1, parent2) and
+    shared_index = [0 .. parent1.getArraySize() - 1] and
+    e1 = parent1.getAnElementExpr(shared_index) and
+    e2 = parent2.getAnElementExpr(shared_index)
   )
 }
 
+// // internal recursive predicate for `hasMultipleInitializerExprsForSameIndex`
```
Comment (knewbury01) [inline c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql:76]: rm if this is not needed (scratch work/alternative version I assume?)

**Review by knewbury01 — CHANGES_REQUESTED** [review]
@lcartey only one small accidental commented out code to remove

otherwise this all looks good to me

File: c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql (hunk 1 above)
Comment (lcartey) [reply c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql]: Done, thanks!

Comment (lcartey) [reply c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql]: It was the previous implementation.

**Review by knewbury01 — APPROVED** [review]
lgtm!

**Pull request merged**